metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py

```diff
@@ -128,17 +128,50 @@ def __init__(self, is_key_schema: bool, default_nullable: bool = False) -> None:
             avro.schema.LogicalSchema: self._gen_non_nested_to_mce_fields,
         }
 
+    @staticmethod
+    def _get_type_name(
+        avro_schema: avro.schema.Schema, logical_if_present: bool = False
+    ) -> str:
+        logical_type_name: Optional[str] = None
+        if logical_if_present:
+            logical_type_name = getattr(
+                avro_schema, "logical_type", None
+            ) or avro_schema.props.get("logicalType")
+        return logical_type_name or str(
+            getattr(avro_schema.type, "type", avro_schema.type)
+        )
+
+    @staticmethod
     def _get_column_type(
-        self, field_type: Union[str, dict], logical_type: Optional[str]
+        avro_schema: avro.schema.Schema, logical_type: Optional[str]
     ) -> SchemaFieldDataType:
-        tp = field_type
-        if hasattr(tp, "type"):
-            tp = tp.type  # type: ignore
-        tp = str(tp)
-        TypeClass: Any = self.field_type_mapping.get(tp)
+        type_name: str = AvroToMceSchemaConverter._get_type_name(avro_schema)
+        TypeClass: Optional[Type] = AvroToMceSchemaConverter.field_type_mapping.get(
+            type_name
+        )
         if logical_type is not None:
-            TypeClass = self.field_logical_type_mapping.get(logical_type, TypeClass)
+            TypeClass = AvroToMceSchemaConverter.field_logical_type_mapping.get(
+                logical_type, TypeClass
+            )
+        assert TypeClass is not None
         dt = SchemaFieldDataType(type=TypeClass())
+        # Handle Arrays and Maps
+        if isinstance(dt.type, ArrayTypeClass) and isinstance(
+            avro_schema, avro.schema.ArraySchema
+        ):
+            dt.type.nestedType = [
+                AvroToMceSchemaConverter._get_type_name(
+                    avro_schema.items, logical_if_present=True
+                )
+            ]
+        elif isinstance(dt.type, MapTypeClass) and isinstance(
+            avro_schema, avro.schema.MapSchema
+        ):
+            # Avro map's key is always a string. See: https://avro.apache.org/docs/current/spec.html#Maps
+            dt.type.keyType = "string"
```

Comment:
That's interesting....  I wonder if DataHub's maps support other key types than `string`?  @rslanka 
I guess this could be added later and be flagged as a known limitation for now?

Reply:
@rslanka Tell me if I get this right, but you are suggesting to use an Avro `array` of `struct` instead of an Avro `map`.  So something like this:

**Current code**
```python
    elif type.is_map_type():
        map_type: IcebergTypes.MapType = type
        kt = _parse_datatype(map_type.key_type())
        vt = _parse_datatype(map_type.value_type())
        # keys are assumed to be strings in avro map
        return {
            "type": "map",
            "values": vt,
            "native_data_type": str(map_type),
            "key_type": kt,
            "key_native_data_type": repr(map_type.key_type()),
        }
```

**Back of napkin new code**
```python
    elif type.is_map_type():
        map_type: IcebergTypes.MapType = type
        kt = _parse_datatype(map_type.key_type())
        vt = _parse_datatype(map_type.value_type())

        return {
            "type": "array",
            "items": {
                "type": "record",
                "fields": [
                    {
                        "name": "key",
                        "type": kt,
                    },
                    {
                        "name": "value",
                        "type": kv,
                    }]
            }
        }
```

If so, the DataHub UI will not display a `Map` type (but a complex record), which would be misleading to the user, no?

Reply:
Looking at the current code as entered here by @cccs-eric, could we just leverage the `key_native_data_type` by getting it from Iceberg for display in the UI?

Reply:
@cccs-eric, yes, your Back of napkin new code looks right in spirit. We can handle this case in the UI later.

@kbendick, the issue is more with converting the ice_berg schema to datahub via the existing AVRO parser. Once it is converted to the native canonicalized datahub schema, UI will simply be using it primarily to avoid having to deal with the different types of source formats that we support.